idioms/ffi-accepting-strings.md

```diff
@@ -0,0 +1,102 @@
+
+The best practice is simple: use `CStr` in such a way as to minimize unsafe code, and create a borrowed slice. If an owned String is needed, call `to_string()` on the string slice.
+
+## Code Example
+
+```rust,ignore
+pub mod unsafe_module {
+
+    // other module content
+
+    #[no_mangle]
+    pub extern "C" fn mylib_log(msg: *const libc::c_char, level: libc::c_int) {
+        let level: crate::LogLevel = match level { /* ... */ };
+
+        let msg_str: &str = unsafe {
+            // SAFETY: accessing raw pointers expected to live for the call, 
+            // and creating a shared reference that does not outlive the current
+            // stack frame.
+            match std::ffi::CStr::from_ptr(msg).to_str() {
+                Ok(s) => s,
+                Err(e) => {
+                    crate::log_error("FFI string conversion failed");
+                    return;
+                }
+            }
+        };
+
+        crate::log(msg_str, level);
+    }
+}
+```
+
+## Advantages
+
+The example is is written to ensure that:
+1. The `unsafe` block is as small as possible.
+2. The pointer with an "untracked" lifetime becomes a "tracked" shared reference
+
+Consider an alternative, where the string is actually copied:
+
+```rust,ignore
+pub mod unsafe_module {
+
+    // other module content
+
+    pub extern "C" fn mylib_log(msg: *const libc::c_char, level: libc::c_int) {
+        /* DO NOT USE THIS CODE. IT IS UGLY, VERBOSE, AND CONTAINS A SUBTLE BUG. */
+
+        let level: crate::LogLevel = match level { /* ... */ };
+
+        let msg_len = unsafe { /* SAFETY: strlen is what it is, I guess? */
+            libc::strlen(msg) + 1 
+        };
+
+        let mut msg_data = Vec::with_capacity(msg_len);
+
+        let msg_cstr: std::ffi::CString = unsafe {
+            // SAFETY: copying from a foreign pointer expected to live 
+            // for the entire stack frame into owned memory
+            std::ptr::copy_nonoverlapping(msg, msg_data.as_mut(), msg_len + 1);
+
+            // SAFETY: setting vector length to copied size, 
+            // so always ininitialized
+            msg_data.set_len(msg_len + 1);
+
+            std::ffi::CString::from_vec_with_nul(msg_data).unwrap()
+        }
+
+        let msg_str: String = unsafe {
+            match msg_cstr.into_string() {
+                Ok(s) => s,
+                Err(e) => {
+                    crate::log_error("FFI string conversion failed");
+                    return;
+                }
+            }
+        };
+
+        crate::log(&msg_str, level);
+    }
+}
+```
+
+This code in inferior to the original in two respects:
+
+1. There is much more `unsafe` code, and more importantly, more invariants it must uphold.
+2. Due to the extensive arithmetic required, there is a bug in this version that cases Rust `undefined behaviour`.
```

Comment:
@jhwgh1968 Last nitpick from my side, I fixed the other suggestions:

Could you please make a small explanation for the bug as you did here?
https://github.com/rust-unofficial/patterns/pull/106/files#diff-e268dfd7a80ac1b0ef9fb2ed7a903397027a5637b1d9d2c84ebb0476e18803dcR147